pkg/controller/resourceclaim/controller.go

```diff
@@ -854,12 +857,52 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 		// those, the resource claim controller will trigger deletion when the
 		// pod is done. However, it doesn't hurt to also trigger deallocation
 		// for such claims and not checking for them keeps this code simpler.
-		if len(valid) == 0 &&
-			claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
-			claim.Status.DeallocationRequested = true
+		if len(valid) == 0 {
+			if builtinControllerFinalizer >= 0 {
+				if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer ||
+					claim.DeletionTimestamp != nil {
+					// Allocated by scheduler with structured parameters. We can "deallocate"
+					// by clearing the allocation.
+					claim.Status.Allocation = nil
+				}
+			} else if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
+				// DRA driver controller in the control plane
+				// needs to do the deallocation.
+				claim.Status.DeallocationRequested = true
+			}
+			// In all other cases, we keep the claim allocated, in particular for immediate allocation
+			// with a control plane controller.
+		}
+
+		claim, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		if err != nil {
+			return err
 		}
 
-		_, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		// Now also remove the finalizer. Note that the index may have changed.
+		builtinControllerFinalizer := slices.Index(claim.Finalizers, resourcev1alpha2.Finalizer)
+		if builtinControllerFinalizer >= 0 {
```

Comment:
don't we need to check if `len(valid) == 0`?

Reply:
Yes, good catch. The E2E testing must be lacking some sanity checks because it has a scenario where the same claim is used by multiple pods.

I'll fix this here first and then later extend the E2E testing to check this.